src/home/homework-blob.c

```diff
@@ -0,0 +1,236 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later
+ *
+ * Copyright © 2024 GNOME Foundation Inc.
+ *      Original Author: Adrian Vovk
+ */
+
+#include "copy.h"
+#include "fileio.h"
+#include "fd-util.h"
+#include "fs-util.h"
+#include "home-util.h"
+#include "homework-blob.h"
+#include "homework.h"
+#include "install-file.h"
+#include "macro.h"
+#include "path-util.h"
+#include "recurse-dir.h"
+#include "rm-rf.h"
+#include "sha256.h"
+#include "string-util.h"
+#include "tmpfile-util.h"
+#include "umask-util.h"
+#include "utf8.h"
+
+#define BLOB_MAX_SIZE (64U*1024U*1024U)
+
+static int copy_one_blob(
+                int src,
+                int dest_dfd,
+                const char *name,
+                uint64_t *total_size,
+                uid_t uid,
+                Hashmap *manifest) {
+        _cleanup_close_ int dest = -EBADF;
+        uint8_t hash[SHA256_DIGEST_SIZE];
+        off_t initial, size;
+        int r;
+
+        if (!hashmap_contains(manifest, name)) {
+                log_warning("File %s in blob directory is missing from manifest. Skipping.", name);
+                return 0;
+        }
+
+        if (!suitable_blob_filename(name)) {
+                log_warning("File %s in blob directory has invalid filename. Skipping.", name);
+                return 0;
+        }
+
+        initial = lseek(src, 0, SEEK_CUR);
+        if (initial < 0)
+                return log_debug_errno(errno, "Failed to get initial pos on fd for %s in blob: %m", name);
+
+        r = sha256_fd(src, hash);
```

Comment:
so, for robustness reasons I'd prefer if sha256_fd() would take a `uint64_t max_size` parameter as third param. If specified as UINT64_MAX the param would have no effect i.e. result in behaviour like currently. But here we'd specify BLOB_MAX_SIZE or so there, which should have the effect that sha256_fd() should simply fail with -EFBIG or so once it reads the max_size+1'st byte. 

i am concerned about sparse files here, which might make us hash forever...